### PR TITLE
feat(ri): add service account auth and closed mode tenant resolution

### PR DESCRIPTION
**Pull request by ashleythedeveloper**
This PR introduces dual tenant mode support to the reference implementation. Operators choose between open mode (one tenant auto-provisioned per user, the existing default) and closed mode (tenants mapped to IdP groups via a configurable JWT claim). Both the session and service account auth flows have been extended to support both modes, forming a 2x2 auth matrix.

## Open mode

Open mode preserves the existing behaviour. Session-based sign-in auto-provisions a tenant per user. Service account Bearer tokens are validated at the middleware, claims propagated as `x-auth-*` headers, and user/tenant auto-provisioned on first API call. A custom NextAuth adapter wrapper prevents duplicate user creation when a service-account-provisioned user later signs in via browser.

> **Note:** Team support (many users sharing one tenant, managed within the application) will be added at a later date.

## Closed mode

Three environment variables control closed mode: `TENANT_MODE=closed`, `TENANT_CLAIM_NAME` (default `groups`), `TENANT_CLAIM_FORMAT` (default `array_first`).

- **Session flow**: `handleSignIn` decodes the access token, extracts the group claim, and creates or links the tenant by `externalIdpGroupId`. `withTenantAuth` resolves the tenant from the session's group claim and re-links users on group change.
- **Service account flow**: `withTenantAuth` validates the Bearer token directly, extracts the group claim, and resolves the tenant via `resolveClosedModeTenant`.
- **Shared resolution**: `resolve-closed-mode-tenant` handles all edge cases (tenant exists/missing x user exists/missing) transactionally with P2002 retry for race conditions.
- **Schema**: Prisma migration adds `externalIdpGroupId` unique column to Tenant.

## E2E test coverage

- Keycloak realm updated with groups, user-group assignments, and protocol mappers
- Docker Compose override (`docker-compose.e2e-closed.yml`) sets closed mode env vars
- Cypress tasks for closed mode cleanup, verification, service account token acquisition, and service account data cleanup
- 4 specs covering the full auth matrix: open x session (existing), open x service account, closed x session, closed x service account

## Test plan

- [x] 911 unit/integration tests passing (new suites: tenant-config, group-claim, keycloak-token, resolve-closed-mode-tenant, with-tenant-auth, adapter-wrapper, handle-sign-in, service-account-user)
- [x] Open mode E2E: session tests pass with no regressions; service account 3/3 passing
- [x] Closed mode E2E: session 7/7 passing (tenant resolution, shared tenant, DID CRUD); service account 4/4 passing
- [x] P2002 race condition retry path covered
- [x] Session auth takes precedence over service account when both are present
- [x] Tokens missing `sub` claim rejected at middleware

### CODE REVIEW

**Comment by ashleythedeveloper** [comment]
The scope of this PR has expanded since the initial Gemini summary above. It now covers dual tenant mode (open/closed). See the updated PR description for the current state.

**Comment by ashleythedeveloper** [comment]
@gemini-code-assist can you review the PR.

File: packages/reference-implementation/src/lib/api/with-tenant-auth.ts [hunk 1]
```diff
@@ -21,41 +30,246 @@ export function withTenantAuth(handler: RouteHandler) {
...
+        'Forbidden — no tenant for group claim',
+      );
+      return NextResponse.json({ ok: false, error: 'No tenant found for group' }, { status: 403 });
+    }
+
+    // Ensure user is linked to the correct tenant (handles group changes)
+    await prisma.user.update({
+      where: { id: session.user.id },
+      data: { tenantId: tenant.id },
+    });
```
[the remark this answers is not in the document]
Comment (ashleythedeveloper) [reply packages/reference-implementation/src/lib/api/with-tenant-auth.ts]: Done in d40e8723 — added findUnique check before update, only writes when tenantId differs.

File: packages/reference-implementation/src/lib/api/with-tenant-auth.ts [hunk 2]
```diff
@@ -21,41 +30,246 @@ export function withTenantAuth(handler: RouteHandler) {
...
+  if (bearerToken) {
+    const validationResult = await validateServiceAccountToken(bearerToken);
+
+    if (!validationResult.valid || !validationResult.payload) {
+      apiLogger.warn(
+        { method, path, error: validationResult.error, durationMs: Date.now() - start },
+        'Unauthorised — invalid bearer token',
+      );
       return NextResponse.json({ ok: false, error: 'Unauthorized' }, { status: 401 });
     }
```
[the remark this answers is not in the document]
Comment (ashleythedeveloper) [reply packages/reference-implementation/src/lib/api/with-tenant-auth.ts:128]: Not addressing — the middleware is intentionally mode-agnostic. It propagates generic claims (sub, azp, name, email) as x-auth-* headers for the open mode path. In closed mode, withTenantAuth needs the full token payload to call extractGroupClaim with the configurable TENANT_CLAIM_NAME. Moving this into the middleware would require it to know about tenant config, breaking the separation of concerns.

File: e2e/cypress/e2e/closed_mode/service-account-api.cy.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    // Fetch a service account token from Keycloak
+    cy.task('getServiceAccountToken').then((result: any) => {
+      accessToken = result.accessToken;
+
+      // Decode the JWT payload to extract the sub claim
+      const payload = JSON.parse(
+        Buffer.from(accessToken.split('.')[1], 'base64').toString(),
+      );
+      tokenSub = payload.sub;
```
[the remark this answers is not in the document]
Comment (ashleythedeveloper) [reply e2e/cypress/e2e/closed_mode/service-account-api.cy.ts]: Done in d40e8723 — removed tokenSub and the JWT decode block.

File: packages/reference-implementation/src/lib/onboarding/handle-sign-in.ts [hunk 4]
```diff
@@ -11,20 +13,124 @@ interface UserProfile {
...
+    await resolveClosedModeSignIn(prisma, userId, groupClaim, account.providerAccountId);
+  } catch (error: unknown) {
+    if (isUniqueConstraintViolation(error)) {
+      logger.warn({ userId, groupClaim }, 'Closed mode sign-in conflict — retrying');
+      await resolveClosedModeSignIn(prisma, userId, groupClaim, account.providerAccountId);
+      return;
+    }
+    throw error;
+  }
+}
```
[the remark this answers is not in the document]
Comment (ashleythedeveloper) [reply packages/reference-implementation/src/lib/onboarding/handle-sign-in.ts:75]: Not addressing — the two functions serve different contexts. handleClosedModeSignIn runs during NextAuth sign-in where the user already exists (created by NextAuth) and needs authProviderId set. resolveClosedModeTenant runs during API calls where the user may not exist and needs full transactional provisioning. The inputs differ (userId + providerAccountId vs sub + claims), the error handling differs, and forcing them through one function would add complexity for both paths.

File: packages/reference-implementation/src/middleware.ts [hunk 5]
```diff
@@ -41,22 +63,51 @@ export default auth(async (req) => {
...
+      // Check Bearer token for service account auth
+      const payload = await validateBearerAuth(req);
+      if (payload) {
+        // sub is required for the service account flow to work
+        if (!payload.sub) {
+          return NextResponse.json(
+            { error: 'Unauthorized', message: 'Token missing required "sub" claim' },
+            { status: 401, headers: { 'x-correlation-id': correlationId } },
+          );
+        }
```
[the remark this answers is not in the document]
Comment (ashleythedeveloper) [reply packages/reference-implementation/src/middleware.ts:85]: Not addressing — the middleware is mode-agnostic by design. It does not have access to getTenantConfig() and does not know whether the app is in open or closed mode. Open mode does not use group claims at all. Adding group claim validation here would require the middleware to be mode-aware, which breaks the separation of concerns. The group claim is validated in withTenantAuth where the mode is known.
